File: ui-browser/src/androidTest/java/l/files/ui/browser/UiBookmarksFragment.java
package l.files.ui.browser;

import androidx.recyclerview.widget.RecyclerView;
import l.files.fs.Path;
import l.files.ui.bookmarks.BookmarksFragment;

import static l.files.base.Objects.requireNonNull;
import static l.files.ui.browser.Instrumentations.*;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

final class UiBookmarksFragment {

    private final UiFileActivity context;

    UiBookmarksFragment(UiFileActivity context) {
        this.context = requireNonNull(context);
    }

    UiFileActivity activityObject() {
        return context;
    }

    UiBookmarksFragment longClick(Path bookmark) {
        longClickItemOnMainThread(
            context.getInstrumentation(),
            this::recycler,
            bookmark
        );
        return this;
    }

    private BookmarksFragment fragment() {
        return (BookmarksFragment) context
            .getActivity()
            .getSupportFragmentManager()
            .findFragmentById(R.id.bookmarks_fragment);
    }

    UiBookmarksFragment click(Path bookmark) {
        assertDrawerIsOpened(true);
        clickItemOnMainThread(
            context.getInstrumentation(),
            this::recycler,
            bookmark
        );
        return this;
    }

    UiBookmarksFragment delete() {
        UiFileActivity activity = activityObject();
        activity.selectActionModeAction(R.id.delete_selected_bookmarks);
        activity.waitForActionModeToFinish();
        return this;
    }

    UiBookmarksFragment assertBookmarked(Path bookmark, boolean bookmarked) {
        awaitOnMainThread(context.getInstrumentation(), () ->
            assertEquals(
                bookmarked,
                fragment().bookmarks().contains(bookmark)
            ));
        return this;
    }

    UiBookmarksFragment assertActionModePresent(boolean present) {
        activityObject().assertActionModePresent(present);
        return this;
    }

    UiBookmarksFragment assertActionModeTitle(Object title) {
        activityObject().assertActionModeTitle(title);
        return this;
    }

    UiBookmarksFragment assertChecked(Path bookmark, boolean checked) {
        findItemOnMainThread(
            context.getInstrumentation(),
            this::recycler,
            bookmark,
            view -> {
                assertEquals(checked, view.isActivated());
                return null;
            }
        );
        return this;
    }

    private RecyclerView recycler() {
        BookmarksFragment fragment = fragment();
        assertNotNull(fragment);
        return fragment.recycler;
    }

    UiBookmarksFragment assertDrawerIsOpened(boolean opened) {
        activityObject().assertDrawerIsOpened(opened);
        return this;
    }

    UiBookmarksFragment pressBack() {
        activityObject().pressBack();
        return this;
    }
}

File: files-app/src/androidTest/kotlin/features/objects/UiBookmarksFragment.kt
package l.files.features.objects

import android.app.Instrumentation
import l.files.fs.Path
import kotlin.test.assertEquals
import android.widget.ListView
import java.util.ArrayList
import kotlin.test.assertTrue
import l.files.ui.bookmarks.BookmarksFragment
import l.files.ui.FilesActivity
import l.files.R

class UiBookmarksFragment(
        private val instrument: Instrumentation,
        private val fragment: BookmarksFragment) {

    private val listView: ListView get() = fragment.getView() as ListView
    private val activity: FilesActivity get() = fragment.getActivity() as FilesActivity

    val activityObject = UiFileActivity(instrument, activity)

    fun selectBookmark(path: Path): UiFileActivity {
        instrument.awaitOnMainThread {
            listView.click { it.equals(path) }
        }
        if (path.resource.readStatus(false).isDirectory) {
            activityObject.assertCurrentDirectory(path)
        }
        return activityObject
    }

    fun checkBookmark(bookmark: Path, checked: Boolean): UiBookmarksFragment {
        instrument.awaitOnMainThread {
            val i = listView.indexOf { it.equals(bookmark) }
            listView.setItemChecked(i, checked)
        }
        return this
    }

    fun deleteCheckedBookmarks(): UiBookmarksFragment {
        activityObject.selectActionModeAction(R.id.delete_selected_bookmarks)
        activityObject.waitForActionModeToFinish()
        return this
    }

    private fun getBookmarks(): List<Path> {
        val paths = ArrayList<Path>()
        for (i in listView.getHeaderViewsCount()..listView.getCount() - 1) {
            paths.add(listView.getItemAtPosition(i) as Path)
        }
        return paths
    }

    fun assertCurrentDirectoryBookmarked(bookmarked: Boolean) = instrument.awaitOnMainThread {
        val path = activity.getCurrentPagerFragment().getCurrentPath()
        val paths = getBookmarks()
        assertEquals(bookmarked, paths.contains(path), paths.toString())
        this
    }

    fun assertBookmarked(bookmark: Path, bookmarked: Boolean) = instrument.awaitOnMainThread {
        assertEquals(bookmarked, getBookmarks().contains(bookmark))
        this
    }

    fun assertContainsBookmarksInOrder(vararg paths: Path): UiBookmarksFragment {
        instrument.awaitOnMainThread {
            assertEquals(paths.toList(), getBookmarks().filter { paths.contains(it) })
        }
        return this;
    }

    private fun ListView.indexOf(pred: (Any) -> Boolean): Int {
        for (i in getHeaderViewsCount()..getCount() - 1) {
            if (pred(getItemAtPosition(i))) {
                return i
            }
        }
        throw AssertionError("Not found")
    }

    private fun ListView.click(pred: (Any) -> Boolean) {
        val position = listView.indexOf(pred)
        val firstVisiblePosition = listView.getFirstVisiblePosition()
        val viewPosition = position - firstVisiblePosition
        val view = listView.getChildAt(viewPosition)
        assertTrue(listView.performItemClick(view, viewPosition, position.toLong()))
    }

}
